src/neo/Network/P2P/Payloads/NotaryRequest.cs

```diff
@@ -0,0 +1,147 @@
+using Neo.IO;
+using Neo.Ledger;
+using Neo.Persistence;
+using Neo.SmartContract;
+using Neo.SmartContract.Native;
+using Neo.VM;
+using System.IO;
+using System.Linq;
+using static Neo.SmartContract.Helper;
+
+namespace Neo.Network.P2P.Payloads
+{
+    public class NotaryRequest : IInventory
+    {
+        /// <summary>
+        /// The transaction need Notary to collect signatures.
+        /// </summary>
+        private Transaction mainTransaction;
+
+        /// <summary>
+        /// This transaction is valid when MainTransaction failed.
+        /// </summary>
+        private Transaction fallbackTransaction;
+
+        /// <summary>
+        /// The witness of the payload. It is a witness of the fallback transaction's signer.
+        /// </summary>
+        private Witness witness;
+        private UInt256 hash = null;
+
+        public InventoryType InventoryType => InventoryType.Notary;
+
+        public UInt256 Hash
+        {
+            get
+            {
+                hash ??= this.CalculateHash();
+                return hash;
+            }
+        }
+
+        public Witness[] Witnesses
+        {
+            get
+            {
+                return new Witness[] { witness };
+            }
+            set
+            {
+                witness = value[0];
+            }
+        }
+
+        public Transaction MainTransaction
+        {
+            get => mainTransaction;
+            set
+            {
+                mainTransaction = value;
+                hash = null;
+            }
+        }
+
+        public Transaction FallbackTransaction
+        {
+            get => fallbackTransaction;
+            set
+            {
+                fallbackTransaction = value;
+                hash = null;
+            }
+        }
+
+        public int Size => mainTransaction.Size + fallbackTransaction.Size + witness.Size;
+
+
+        public void DeserializeUnsigned(BinaryReader reader)
+        {
+            mainTransaction = reader.ReadSerializable<Transaction>();
+            fallbackTransaction = reader.ReadSerializable<Transaction>();
+        }
+
+        public void Deserialize(BinaryReader reader)
+        {
+            DeserializeUnsigned(reader);
+            witness = reader.ReadSerializable<Witness>();
+        }
+
+        public void Serialize(BinaryWriter writer)
+        {
+            SerializeUnsigned(writer);
+            writer.Write(witness);
+        }
+
+        public void SerializeUnsigned(BinaryWriter writer)
+        {
+            writer.Write(mainTransaction);
+            writer.Write(fallbackTransaction);
+        }
+
+        public UInt160[] GetScriptHashesForVerifying(DataCache snapshot)
+        {
+            return new UInt160[] { fallbackTransaction.Signers[1].Account };
```

Comment:
what if teh Signers size is 1 here?

Reply:
Fallback tx MUST have 2 signer. The check is here https://github.com/neo-ngd/neo/blob/4b10f64835eade8edb2005d9171cd2b839639f9a/src/neo/Network/P2P/Payloads/NotaryRequest.cs#L108